src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/SendOrderContestedAboutToSubmitHandler.java

```diff
@@ -55,139 +56,302 @@ public boolean canHandle(CallbackType callbackType, CaseType caseType, EventType
+            shareAndSendGeneralOrderWithSelectedParties(caseDetails, parties, selectedOrders, printOrderCollection);
+
+
+            log.info("Share and print hearing order for case {}", caseDetails.getId());
+            List<CaseDocument> hearingOrders = generalOrderService.hearingOrdersToShare(caseDetails, selectedOrders);
+            if (hearingOrders != null && !hearingOrders.isEmpty()) {
+                shareAndSendHearingDocuments(caseDetails, hearingOrders, parties, printOrderCollection, userAuthorisation);
+                log.info("sending for stamp final order on case {}", caseDetails.getId());
+                hearingOrders.forEach(orderToStamp -> {
+                    log.info("StampFinalOrder {} for Case ID {}, ", orderToStamp, caseId);
+                    stampAndAddToCollection(caseDetails, orderToStamp, userAuthorisation);
+                });
+            }
+            caseData.setOrdersSentToPartiesCollection(printOrderCollection);
+        } catch (RuntimeException e) {
+            return GenericAboutToStartOrSubmitCallbackResponse.<FinremCaseData>builder()
+                .data(caseDetails.getData()).errors(List.of(e.getMessage())).build();
         }
 
-        return GenericAboutToStartOrSubmitCallbackResponse.<Map<String, Object>>builder().data(caseDetails.getData()).build();
+        return GenericAboutToStartOrSubmitCallbackResponse.<FinremCaseData>builder()
+            .data(caseDetails.getData()).build();
     }
 
-    private void stampFinalOrder(CaseDetails caseDetails, String authToken) {
-        Map<String, Object> caseData = caseDetails.getData();
-
-        List<HearingOrderCollectionData> hearingOrderCollectionData = documentHelper.getHearingOrderDocuments(caseData);
 
-        if (hearingOrderCollectionData != null && !hearingOrderCollectionData.isEmpty()) {
-            int index = hearingOrderCollectionData.size() - 1;
-            CaseDocument latestHearingOrder = hearingOrderCollectionData
-                .get(index)
-                .getHearingOrderDocuments().getUploadDraftDocument();
+    private void shareAndSendHearingDocuments(FinremCaseDetails caseDetails,
+                                              List<CaseDocument> hearingOrders,
+                                              List<String> partyList,
+                                              List<OrderSentToPartiesCollection> printOrderCollection,
+                                              String userAuthorisation) {
+        Long caseId = caseDetails.getId();
+        log.info("Share Hearing Documents for case {}:", caseId);
+        List<CaseDocument> hearingDocumentPack = createHearingDocumentPack(caseDetails, hearingOrders, userAuthorisation);
+        hearingDocumentPack.forEach(doc -> printOrderCollection.add(addToPrintOrderCollection(doc)));
+
+        FinremCaseData caseData = caseDetails.getData();
+        if (partyList.contains(CaseRole.APP_SOLICITOR.getCcdCode())) {
```

Comment:
Can we refactor the code that sets the order collection for each party as there seems to be a lot of repetition here

Reply:
You have done this.  Thank you.